python/ql/src/experimental/Security-new-dataflow/CWE-078/CommandInjection.ql

```diff
@@ -27,19 +27,30 @@ class CommandInjectionConfiguration extends TaintTracking::Configuration {
   override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
 
   override predicate isSink(DataFlow::Node sink) {
-    sink = any(SystemCommandExecution e).getCommand()
+    sink = any(SystemCommandExecution e).getCommand() and
+    // Since the implementation of os.popen looks like
+    // ```py
+    // def popen(cmd, mode="r", buffering=-1):
+    //     ...
+    //     proc = subprocess.Popen(cmd, ...)
+    // ```
+    // any time we would report flow to the `os.popen` sink, we can ALSO report the flow
+    // from the `cmd` parameter to the `subprocess.Popen` sink -- obviously we don't want
+    // that.
+    //
+    // However, simply removing taint edges out of a sink is not a good enough solution,
+    // since we would only flag one of the `os.system` calls in the following example due
+    // to use-use flow
+    // ```py
+    // os.system(cmd)
+    // os.system(cmd)
+    // ```
+    //
+    // Best solution I could come up with is to exclude all sinks inside the standard
+    // library -- this does have a downside: If we have overlooked a function in the
+    // standard library that internally runs a command, we no longer give an alert :|
+    not sink.getLocation().getFile().inStdlib()
```

Comment:
Maybe it would be better to _not_ do this, and to explicitly exclude the sink inside `os.popen` for that particular `SystemCommandExecution`?

Reply:
If it only was so simple :disappointed: I knew this would also be a problem for `subprocess.call` and such helpers, but I realize this comment makes it sound like it is ONLY about `os.popen`.

Anyway, after excluding flow for `os.system`, `os.popen`, `subprocess.call`, and so on, I realized that I didn't think about the `os.exec*` functions. [Lo and behold](https://github.com/python/cpython/blob/fa7ce080175f65d678a7d5756c94f82887fc9803/Lib/os.py#L537-L542), they would also fall into this category... but I would easily have forgotten about them :| Based on this discovery it would be really neat with a general approach to this.

I'll ask around to see if other language teams have a good solution to this.

Reply:
On second thought, disregarding sinks that appear inside the standard library seems somewhat reasonable to me. 

Perhaps if we had _some_ way of figuring out which functions internally run a particular command, we could use this to minimise the risk of the downside you list above... :slightly_smiling_face:

Reply:
discussed in person, we'll stick with only excluding sinks inside `os` and `subprocess` instead of the _whole_ standard library.